===================================================================
RCS file: /cvs/glibc/libc/catgets/catgets.c,v
retrieving revision 1.14.2.1
retrieving revision 1.14.2.2
diff -u -r1.14.2.1 -r1.14.2.2
--- libc/catgets/catgets.c	2000/01/29 11:52:12	1.14.2.1
+++ libc/catgets/catgets.c	2000/08/21 21:01:52	1.14.2.2
@@ -50,7 +50,9 @@
 	/* Use the LANG environment variable.  */
 	env_var = getenv ("LANG");
 
-      if (env_var == NULL)
+
+      if (env_var == NULL || *env_var == '\0'
+	  || (__libc_enable_secure && strchr (env_var, '/') != NULL))
 	env_var = "C";
 
       env_var_len = strlen (env_var) + 1;
===================================================================
RCS file: /cvs/glibc/libc/locale/findlocale.c,v
retrieving revision 1.10.2.1
retrieving revision 1.10.2.4
diff -u -r1.10.2.1 -r1.10.2.4
--- libc/locale/findlocale.c	1999/11/08 23:45:13	1.10.2.1
+++ libc/locale/findlocale.c	2000/08/27 19:00:32	1.10.2.4
@@ -1,4 +1,4 @@
-/* Copyright (C) 1996, 1997, 1998, 1999 Free Software Foundation, Inc.
+/* Copyright (C) 1996, 1997, 1998, 1999, 2000 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
    Contributed by Ulrich Drepper <drepper@gnu.org>, 1996.
 
@@ -54,11 +54,7 @@
   const char *revision;
   struct loaded_l10nfile *locale_file;
 
-  if ((*name)[0] == '\0'
-      /* In SUID binaries we must not allow people to access files
-	 outside the dedicated locale directories.  */
-      || (__libc_enable_secure
-	  && memchr (*name, '/', _nl_find_language (*name) - *name) != NULL))
+  if ((*name)[0] == '\0')
     {
       /* The user decides which locale to use by setting environment
 	 variables.  */
@@ -67,9 +63,12 @@
 	*name = getenv (_nl_category_names[category]);
       if (*name == NULL || (*name)[0] == '\0')
 	*name = getenv ("LANG");
-      if (*name == NULL || (*name)[0] == '\0')
-	*name = (char *) _nl_C_name;
     }
+
+  if (*name == NULL || (*name)[0] == '\0'
+      || (__builtin_expect (__libc_enable_secure, 0)
+	  && strchr (*name, '/') != NULL))
+    *name = (char *) _nl_C_name;
 
   if (strcmp (*name, _nl_C_name) == 0 || strcmp (*name, _nl_POSIX_name) == 0)
     {
===================================================================
RCS file: /cvs/glibc/libc/intl/dcgettext.c,v
retrieving revision 1.26
retrieving revision 1.26.2.2
diff -u -r1.26 -r1.26.2.2
--- libc/intl/dcgettext.c	1999/02/09 08:26:37	1.26
+++ libc/intl/dcgettext.c	2000/08/27 19:01:29	1.26.2.2
@@ -1,5 +1,5 @@
 /* Implementation of the dcgettext(3) function.
-   Copyright (C) 1995, 1996, 1997, 1998, 1999 Free Software Foundation, Inc.
+   Copyright (C) 1995,1996,1997,1998,1999,2000 Free Software Foundation, Inc.
 
    This file is part of the GNU C Library.  Its master source is NOT part of
    the C library, however.
@@ -371,10 +371,7 @@
 
 	  /* When this is a SUID binary we must not allow accessing files
 	     outside the dedicated directories.  */
-	  if (ENABLE_SECURE
-	      && (memchr (single_locale, '/',
-			  _nl_find_language (single_locale) - single_locale)
-		  != NULL))
+	  if (ENABLE_SECURE && strchr (single_locale, '/') != NULL)
 	    /* Ingore this entry.  */
 	    continue;
 	}
